evm/src/generation/prover_input.rs

```diff
@@ -123,6 +127,93 @@ impl<F: Field> GenerationState<F> {
+            let a_start_loc = stack_peek(self, 2)
+                .expect("Stack does not have enough items")
+                .try_into()
+                .unwrap();
+            let b_start_loc = stack_peek(self, 3)
+                .expect("Stack does not have enough items")
+                .try_into()
+                .unwrap();
+            let m_start_loc = stack_peek(self, 4)
+                .expect("Stack does not have enough items")
+                .try_into()
+                .unwrap();
+
+            let result = match function {
+                "remainder" => {
+                    self.bignum_modmul_remainder(len, a_start_loc, b_start_loc, m_start_loc)
+                }
+                "quotient" => {
+                    self.bignum_modmul_quotient(len, a_start_loc, b_start_loc, m_start_loc)
+                }
+                _ => panic!("Invalid prover input function."),
+            };
+
+            self.bignum_modmul_prover_inputs = result
+                .iter()
+                .cloned()
+                .pad_using(len, |_| 0.into())
+                .collect();
+            self.bignum_modmul_prover_inputs.reverse();
+        }
+
+        self.bignum_modmul_prover_inputs.pop().unwrap()
+    }
+
+    fn bignum_modmul_remainder(
+        &mut self,
+        len: usize,
+        a_start_loc: usize,
+        b_start_loc: usize,
+        m_start_loc: usize,
+    ) -> Vec<U256> {
+        let a = &self.memory.contexts[0].segments[Segment::KernelGeneral as usize].content
+            [a_start_loc..a_start_loc + len];
+        let b = &self.memory.contexts[0].segments[Segment::KernelGeneral as usize].content
+            [b_start_loc..b_start_loc + len];
+        let m = &self.memory.contexts[0].segments[Segment::KernelGeneral as usize].content
+            [m_start_loc..m_start_loc + len];
+
+        let a_biguint = mem_vec_to_biguint(a);
+        let b_biguint = mem_vec_to_biguint(b);
+        let m_biguint = mem_vec_to_biguint(m);
+
+        let result_biguint = (a_biguint * b_biguint) % m_biguint;
+        biguint_to_mem_vec(result_biguint)
+    }
+
+    fn bignum_modmul_quotient(
+        &mut self,
+        len: usize,
+        a_start_loc: usize,
+        b_start_loc: usize,
+        m_start_loc: usize,
+    ) -> Vec<U256> {
+        let a = &self.memory.contexts[0].segments[Segment::KernelGeneral as usize].content
+            [a_start_loc..a_start_loc + len];
+        let b = &self.memory.contexts[0].segments[Segment::KernelGeneral as usize].content
+            [b_start_loc..b_start_loc + len];
+        let m = &self.memory.contexts[0].segments[Segment::KernelGeneral as usize].content
+            [m_start_loc..m_start_loc + len];
+
+        let a_biguint = mem_vec_to_biguint(a);
+        let b_biguint = mem_vec_to_biguint(b);
+        let m_biguint = mem_vec_to_biguint(m);
+
+        let result_biguint = (a_biguint * b_biguint) / m_biguint;
+        biguint_to_mem_vec(result_biguint)
+    }
```

Comment:
These should ideally be one function, since they are virtually the same—division and remainder are in some sense the same operation (once you have one, you can get the other one quite cheaply).

I would calculate
```rust
let prod = a_biguint * b_biguint;
let quo = prod / m_biguint;
let rem = prod - quo * m_biguint;
```

I think this would be cleaner code, in addition to being faster (because you don't repeat the same division twice).

(Ideally `num_bigint` would have a `divmod` function, but I couldn't find one…)